### ISSUE TITLE
What would be needed for an array-based version of Itertools::tuples()?

### ISSUE THREAD

**Issue by HadrienG2** [bluss/rust-itertools#243]
Itertools::tuples() is a bit of a strange beast. Since Rust iterators are guaranteed to always output the same type, it would make more sense to group consecutive items in arrays, as opposed to tuples. Do we agree on this? And if so, what would be needed to make it happen?

**Comment by bluss** [comment]
Tuples support destructing in let, for and elsewhere and arrays do not and that's pretty much it. It is a practical decision. I'd require both array patterns and const generics to use arrays.

**Comment by HadrienG2** [comment]
Array patterns [are already available in nightly](https://play.rust-lang.org/?gist=7a5df535ef9488dfcf497f0b66347005&version=nightly), but I can understand why you would want to wait for const generics as well. Otherwise, we would end up replacing one imperfect solution with another.

Let's wait a bit, then...

**Comment by bluss** [comment]
unstable features doesn't matter to itertools, we'd need them to be stable of course. Having slice patterns would be exciting, and long overdue.

**Comment by HadrienG2** [comment]
As slice patterns are stable since 1.26, maybe this issue could be revisited at the next Rust version bump?

**Comment by amatveiakin** [comment]
Const generics have been stabilized. So [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2a8d495d3b49d445a5679465b7f7c333) is possible now:

```rust
fn get_ints<const N: usize>() -> [u32; N] {
    [(); N].map(|()| rand::random())
}

let [a, b, c] = get_ints();
```

I believe it’s time to add array versions of all tuple functions: `tuples`, `tuple_windows`, `collect_tuple`, etc. WDYT?

I would be happy to make a pull request.

**Comment by jswrenn** [comment]
I agree it's time to do this, and we already have several PRs pending for it.

However, now the blocker isn't MSRV, but rather dependencies. At the time those PRs were filed, Rust was still missing key helpers for safely initializing arrays. The unsafe code needed for this is VERY subtle, and I do not want us to try to do it on our own. Those PRs worked around that by adding a dependency for array initialization, but itertools also takes a very conservative stance on dependencies.

The next step for this is to look at those PRs, see what routines can be replaced with standard library helpers, or else gate these array features (and the associated dependency) behind a cargo feature flag.

**Comment by HadrienG2** [comment]
Note that [`core::array::from_fn()`](https://doc.rust-lang.org/stable/core/array/fn.from_fn.html) can be used for many different array initialization scenarios, and generates pretty decent code on modern rustc versions.

**Comment by amatveiakin** [comment]
With [`core::array::from_fn`](https://doc.rust-lang.org/stable/core/array/fn.from_fn.html) it's easy to write a function that returns an array or panics. In order to return an `Option<[T; N]>` we would need to first construct an `[Option<T>; N]`, right? So if the goal is to avoid both unsafe code and performance overhead, then we need [`core::array::try_from_fn`](https://doc.rust-lang.org/stable/core/array/fn.try_from_fn.html), which is not stable yet.
